qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,208 @@
+        """Compute frequencies/counts of phases from the result of running the QPE circuit.
+
+        How the frequencies are computed depends on whether the backend computes amplitude or
+        samples outcomes.
+
+        1) If the backend is a statevector simulator, then the reduced density matrix of the
+        phase-reading register is computed from the combined phase-reading- and input-state
+        registers. The elements of the diagonal `(i, i)` give the probability to measure the each of
+        the states `i`. The index `i` expressed as a binary integer with the LSB rightmost gives the
+        state of the phase-reading register with the LSB leftmost when interpreted as a phase. In
+        order to maintain the compact representation, the phases are maintained as decimal integers.
+        They may be converted to other forms via the results object, `PhaseEstimatorResult` or
+        `HamiltonianPEResult`.
+
+         2) If the backend samples bitstrings, then the counts are first retrieved as a dict.  The
+        binary strings (the keys) are then reversed so that the LSB is rightmost and the counts are
+        converted to frequencies. Then the keys are sorted according to increasing phase, so that
+        they can be easily understood when displaying or plotting a histogram.
+
+        Args:
+            circuit_result: the result object returned by the backend that ran the QPE circuit.
+
+        Returns:
+               Either a dict or numpy.ndarray representing the frequencies of the phases.
+
+        """
+        if self._quantum_instance.is_statevector:
+            state_vec = circuit_result.get_statevector()
+            evaluation_density_matrix = get_subsystem_density_matrix(
+                state_vec,
+                range(self._num_evaluation_qubits,
+                      self._num_evaluation_qubits + self._num_unitary_qubits)
+            )
+            # The diagonal of a density matrix is real, so imaginary part should be zero.
+            phases = evaluation_density_matrix.diagonal().real
+        else:
+            # return counts with keys sorted numerically
+            num_shots = circuit_result.results[0].shots
+            counts = circuit_result.get_counts()
+            phases = {k[::-1]: counts[k] / num_shots for k in counts.keys()}
+            phases = _sort_phases(phases)
+            phases = qiskit.result.Counts(phases, memory_slots=counts.memory_slots,
+                                          creg_sizes=counts.creg_sizes)
+
+        return phases
+
+    def _run(self) -> PhaseEstimatorResult:
+        """Run the circuit and return and return `PhaseEstimatorResult`.
+
+
+        Returns:
+               An instance of qiskit.algorithms.phase_estimator_result.PhaseEstimatorResult.
+        """
+
+        circuit_result = self._quantum_instance.execute(self.construct_circuit())
+        phases = self._compute_phases(circuit_result)
+        if isinstance(phases, numpy.ndarray):
```

Comment:
Would it make sense to have the results object have a case distinction for `phases` being an array or a dict? Then we wouldn't have to have this distinction here.

Concretely, I'm suggesting (in pseudo-code)
```python
class PhaseEstimationResult(AlgorithmResult):
    def __init__(self, ..., phases: Union[dict, array]):
        ...
    
    # wherever phases is used just have
    if isinstance(self.phases, dict): 
         # handle as phase_dict
    else:
        # handle as phase_array
```
I think this is a bit cleaner and more Pythonic than multiple arguments just for different types.

Reply:
Good idea. The change has been made.